Add deployments oidc pages (#2750)
buildkite/docs: pages/deployments/oidc/oidc_with_aws.md

---
keywords: oidc, authentication, IAM, roles, AWS
---

# OIDC with AWS

> The [Buildkite Agent's oidc command](/docs/agent/v3/cli-oidc) allows you to request an OIDC token representing the current job. These tokens can be exchanged with federated systems like AWS in exchange for an IAM role with AWS scoped permissions.

Reviewer: Need to initially spell out what IAM is (i.e. "Identity and Access Management (IAM)").

> The [Buildkite Agent's oidc command](/docs/agent/v3/cli-oidc) allows you to request an OIDC token representing the current job. These tokens can be exchanged on federated systems like AWS for an identity and access management (IAM) role with AWS-scoped permissions.

Author: ```suggestion
The [Buildkite Agent's oidc command](/docs/agent/v3/cli-oidc) allows you to request an OIDC token representing the current job. These tokens can be exchanged on federated systems like AWS for an Identity and Access Management (IAM) role with AWS-scoped permissions.
```

Reviewer: Looking at the AWS docs, it looks like IAM is a product rather than feature name (at least that's how Amazon is conveying this in their docs, even though Amazon don't appear to offer clear style guidance on distinguishing the two). Therefore, let's follow suit in our docs.

See the [OpenID Connect Core documentation](https://openid.net/specs/openid-connect-core-1_0.html#IDToken) for more information about how OIDC tokens are constructed and how to extract and use claims.

See the [AWS OpenID Connect identity provider in IAM documentation](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_create_oidc.html) for more information on their implementation of the federated system.

We will be using two Buildkite Plugins to utilise OIDC with AWS and our Pipelines.

- Buildkite Plugin [aws-assume-role-with-web-identity-buildkite-plugin](https://github.com/buildkite-plugins/aws-assume-role-with-web-identity-buildkite-plugin)
- Buildkite Plugin [aws-ssm-buildkite-plugin](https://github.com/buildkite-plugins/aws-ssm-buildkite-plugin)

## Step 1: Setup OIDC provider in AWS account

First we'll set up an AWS IAM OpenID Connect (OIDC) provider using the following configuration.

  URL: `https://agent.buildkite.com`

>   Audience: `sts.amazonaws.com`

Author: Originally I had a screenshot in here, but I think we could instead link to the AWS documentation on _how_ to create an AWS IAM OpenID Connect provider https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_create_oidc.html

Reviewer: For tasks that are conducted on 3rd party systems, it's usually good/best practice to link through to the relevant documentation for those systems, since these 3rd parties are in control of developing these features and updating their documentation for it.

If we (re-)document such 3rd-party content in our docs, we then become responsible for updating this content whenever these 3rd parties update their product functionality.

The exception to this would be where specific steps might require specific tasks or Buildkite-specific code configurations to be specified. Therefore, it'd be better to reference the 3rd party process and make reference in our docs to things like, for example: "At step 3, copy and paste the following code snippet and modify accordingly, etc."

Author: Sweet, thanks, then that URL and Audience information is the only part they need to copy and paste, and we should link to that AWS documentation page.

> ## Step 2: Create a new (or update an existing) IAM role to use with your pipeline(s)

Author: Any maybe in here we should also link to the AWS docs on creating an IAM role that works with OpenID connect 

https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_create_for-idp_oidc.html

> In this example the IAM role is named: `compute-ssm-oidc-example`. When creating the role you must `Select Custom Trust Policy`.

Author: Any maybe in here we should also link to the AWS docs on creating an IAM role with a custom trust policy   https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_create_for-custom.html

> Set the Principal to be federated via the IAM OIDC provider we just created, and add some conditions against using this IAM role for additional security using the format of the OIDC token subject provided by Buildkite.

Reviewer: What exactly is the "Principal" and whereabouts is this configured?

            },
            "Action": "sts:AssumeRoleWithWebIdentity",
            "Condition": {
                "StringEquals": {
                    "agent.buildkite.com:aud": "sts.amazonaws.com"
                },
                "StringLike": {
                    "agent.buildkite.com:sub": "organization\:ORGANIZATION_SLUG\:pipeline:PIPELINE_SLUG\:ref\:REF\:commit\:BUILD_COMMIT\:step\:STEP_KEY"
                },
                "IpAddress": {
                    "aws:SourceIp": [
                        "AGENT_PUBLIC_IP_ONE",
                        "AGENT_PUBLIC_IP_TWO"
                    ]
                }

            }
        }
    ]
}
```

Update the following sections:

- In the Principal, replace `AWS_ACCOUNT_ID`  with your account ID.
- In the Condition on the subject, replace:

    * `ORGANIZATION_SLUG` with your Buildkite Organization
    * `PIPELINE_SLUG` with your Pipeline
    * `REF` - this is commonly replaced with `refs/heads/main` to enforce only the main branch using the IAM role  or `refs/tags/*` for only tagged releases able to deploy, or a wildcard if we want all branches to use it.
    * `BUILD_COMMIT` - this is commonly replaced with a wildcard `*`

>     * `STEP_KEY` - this is commonly replaced with a wildcard `*`

Author: There might be a nicer way to format "replace these parts of the code snippet above" than what I've put in here, but I leave that to your expert opinion!

Reviewer: Great thanks for this :-) I've reworked this somewhat—you'll see in my next commit, but it'd be ideal if you could answer my remaining queries below.

Expanded example Trust Policy:

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Principal": {
                "Federated": "arn\:aws\:iam:\:AWS_ACCOUNT_ID\:oidc-provider/agent.buildkite.com"
            },
            "Action": "sts:AssumeRoleWithWebIdentity",
            "Condition": {
                "StringEquals": {
                    "agent.buildkite.com:aud": "sts.amazonaws.com"
                },
                "StringLike": {
                    "agent.buildkite.com:sub": [
                        "organization\:example-org\:pipeline\:example-pipeline\:ref\:refs/heads/main\:*",
                        "organization\:example-org\:pipeline\:example-pipeline\:ref\:refs/tags/*:*"
                    ]
                },
                "IpAddress": {
                    "aws:SourceIp": [
                        "192.0.2.0",
                        "198.51.100.0"
                    ]
                }

            }
        }
    ]
}
```

## Step 3: Configure your IAM role with AWS actions

> Add inline or managed IAM policies to the role to allow whatever IAM permissions your pipeline needs. Common examples are permissions to read secrets from SSM and push images to ECR, but there are no limitations here.

Author: These are the AWS docs for the inline or managed  IAM policies that need to be attached to the role to set it up to interact with the customers AWS account. 

https://docs.aws.amazon.com/IAM/latest/UserGuide/access_policies_managed-vs-inline.html

```suggestion
Add an inline or managed IAM policy to the role to allow the IAM role to perform any actions your pipeline needs. Common examples are permissions to read secrets from SSM and push images to ECR, but this entirely depends on the purpose of your Pipeline.
```

I didn't add the link in there, but that's slightly better phrasing of what this step is doing

## Step 1: Set up an OIDC provider in your AWS account

First, you'll need to set up an IAM OIDC provider in your AWS account.

Learn more about how to do this in the [Create an OpenID Connect (OIDC) identity provider in IAM](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_create_oidc.html) page of the AWS documentation.

On this page, as part of the [Creating and managing an OIDC provider (console)](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_create_oidc.html#manage-oidc-provider-console) process, specify the following values for the:

- **Provider URL**: `https://agent.buildkite.com`

- **Audience**: `sts.amazonaws.com`

## Step 2: Create a new (or update an existing) IAM role to use with your pipelines

Creating new or updating existing IAM roles is conducted through your AWS account.

Learn more about how to do this in the [Creating a role using custom trust policies (console)](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_create_for-custom.html) page of the AWS documentation.

As part of this process:

1. Choose the **Custom trust policy** role type.

> 1. Copy the following example trust policy in the following JSON code block and paste it into an code editor:

Author: ```suggestion
1. Copy the following example trust policy in the following JSON code block and paste it into a code editor:
```

Reviewer: Whoops 🤭

    ```json
    {
        "Version": "2012-10-17",
        "Statement": [
            {
                "Effect": "Allow",
                "Principal": {

>                     "Federated": "arn\:aws\:iam:\:AWS_ACCOUNT_ID\:oidc-provider/agent.buildkite.com"

Reviewer: Other than `AWS_ACCOUNT_ID`, is anything else in this string value likely to change?

Author: Nothing else should change - it's the ARN of the OIDC provider they created in the previous step, and because the provider URL is the one we specify, then I don't think it's possible for it to be named anything different,

                    },
                    "IpAddress": {
                        "aws:SourceIp": [
                            "AGENT_PUBLIC_IP_ONE",
                            "AGENT_PUBLIC_IP_TWO"
                        ]
                    }

                }
            }
        ]
    }
    ```

1. Modify the following sections of the pasted code snippet accordingly:
    * `AWS_ACCOUNT_ID` (in `Principal`) is your AWS account ID.
    * next item

1. In the **Custom trust policy** section,

1. Specify an appropriate **Role name**, for example, `compute-ssm-oidc-example`.

Set the Principal to be federated via the IAM OIDC provider we just created, and add some conditions against using this IAM role for additional security using the format of the OIDC token subject provided by Buildkite.

This allows us to limit use of this IAM role to named Organizations, Pipelines, and optionally branches, commits and steps. Format is
 `organization:ORGANIZATION_SLUG:pipeline:PIPELINE_SLUG:ref:REF:commit:BUILD_COMMIT:step:STEP_KEY`

You can use a wildcard `*` to replace any subjects you don’t want to set limits on.

> You can also add multiple token subjects as a list if you want to use the same IAM role for multiple pipelines, or if you want a variety of branches, commits and steps enabled.

Author: So this is what I did in the expanded example at the bottom, but there's basically infinite combinations of things we can do here. 

This is Buildkite organisation `example-org` and their pipeline `example-pipeline`, which can ONLY be exchanged for an IAM role when building on the main branch and tagged releases. 
```
"StringLike": {
                     "agent.buildkite.com:sub": [
                         "organization:example-org:pipeline:example-pipeline:ref:refs/heads/main:*",
                         "organization:example-org:pipeline:example-pipeline:ref:refs/tags/*:*"
                     ]
                 }
```

This is Buildkite organisation `example-org` and their pipeline `example-pipeline`, which can be exchanged for an IAM role when building on ANY branch or tag. 
```
"StringLike": {
                     "agent.buildkite.com:sub": [
                         "organization:example-org:pipeline:example-pipeline:ref:refs:*",
                     ]
                 }
```

This is Buildkite organisation `example-org` and multiple pipelines pipeline `example-pipeline-one`, `example-pipeline-two` and `example-pipeline-three`, which can each use this role when building on any branch or tag. 
```
"StringLike": {
                     "agent.buildkite.com:sub": [
                         "organization:example-org:pipeline:example-pipeline-one:ref:refs:*",
                         "organization:example-org:pipeline:example-pipeline-two:ref:refs:*",
                         "organization:example-org:pipeline:example-pipeline-three:ref:refs:*",
                     ]
                 }
```

This is Buildkite organisation `example-org` and allowing any pipeline in their org, which can use this role when building on any branch or tag. 
```
"StringLike": {
                     "agent.buildkite.com:sub": [
                         "organization:example-org:pipeline:*:ref:refs:*",
                     ]
                 }
```

> You can also add multiple token subjects as a list if you want to use the same IAM role for multiple pipelines, or if you want a variety of branches, commits and steps enabled.

Author: This is the write up on why that last example wasn't previous secure. We've mitigated the vulnerability so it shouldn't be possible, but maybe let's not recommend that just in case. https://3.basecamp.com/3453178/buckets/3258372/messages/7155932480

> You can also add multiple token subjects as a list if you want to use the same IAM role for multiple pipelines, or if you want a variety of branches, commits and steps enabled.

Reviewer: Thanks, really useful to know, and I hope I've re-written the instructions to capture this information

> If you have a public IP address associated with your Buildkite Agent's, you can also set a condition on the source IP address and further restrict access to your IAM role.

Reviewer: Does this relate to the:

```
                    "IpAddress": {
                        "aws:SourceIp": [
                            "AGENT_PUBLIC_IP_ONE",
                            "AGENT_PUBLIC_IP_TWO"
                        ]
                    }
```

part of the trust policy (JSON) code snippet?

Author: Yes that's right. We can say that only token exchange requests coming from those IP addresses are valid, and it'll ignore requests from anywhere else.

Customers don't have to assign IP addresses to their agents, and in the new hosted agents they _cannot_ assign IP addresses to their agents, which is why it's being talked about as an aside.

It's a really useful option for folks to use though, and we use it extensively internally.

Reviewer: Useful to know! Thanks, I'll put some of this key info into the first cut of these docs.

Update the following sections:

- In the Condition on the subject, replace:

    * `ORGANIZATION_SLUG` with your Buildkite Organization
    * `PIPELINE_SLUG` with your Pipeline
    * `REF` - this is commonly replaced with `refs/heads/main` to enforce only the main branch using the IAM role  or `refs/tags/*` for only tagged releases able to deploy, or a wildcard if we want all branches to use it.
    * `BUILD_COMMIT` - this is commonly replaced with a wildcard `*`

>     * `STEP_KEY` - this is commonly replaced with a wildcard `*`

Reviewer: From your example below, it looks like these two values are optional.
Are these definitely optional values and if so, what value is assumed by omitting them in the trust policy?

Author: They are definitely optional, but you have to end the string in a wildcard `*` if you do, e.g. 
```
"StringLike": {
                     "agent.buildkite.com:sub": [
                         "organization:example-org:pipeline:*:ref:refs:*",
                     ]
                 }
```

Author: That whole string is based on the format we set when we issue the token from the agent - see our docs here https://buildkite.com/docs/agent/v3/cli-oidc 
<img width="845" alt="Screenshot 2024-04-30 at 2 54 09 PM" src="https://github.com/buildkite/docs/assets/23623393/fc322102-8a37-416c-a138-8219ac39967f">

We can match to every single aspect of that string, or just parts on it, depending on how restrictive we want to be.

Reviewer: Hmm, thanks for this additional info, I might make a note of this in my next commit to these docs.

---
keywords: oidc, authentication, IAM, roles, AWS
---

# OIDC with AWS

> The [Buildkite Agent's oidc command](/docs/agent/v3/cli-oidc) allows you to request an OpenID Connect (OIDC) token representing the current job. These tokens can be exchanged on federated systems like AWS for an Identity and Access Management (IAM) role with AWS-scoped permissions.

Reviewer: Good point!

1. Copy the following example trust policy in the following JSON code block and paste it into a code editor:

    ```json
    {
        "Version": "2012-10-17",
        "Statement": [
            {
                "Effect": "Allow",
                "Principal": {
                    "Federated": "arn\:aws\:iam:\:AWS_ACCOUNT_ID\:oidc-provider/agent.buildkite.com"
                },
                "Action": "sts:AssumeRoleWithWebIdentity",
                "Condition": {
                    "StringEquals": {
                        "agent.buildkite.com:aud": "sts.amazonaws.com"
                    },
                    "StringLike": {
                        "agent.buildkite.com:sub": "organization\:ORGANIZATION_SLUG\:pipeline:PIPELINE_SLUG\:ref\:REF\:commit\:BUILD_COMMIT\:step\:STEP_KEY"
                    },
                    "IpAddress": {
                        "aws:SourceIp": [
                            "AGENT_PUBLIC_IP_ONE",
                            "AGENT_PUBLIC_IP_TWO"
                        ]
                    }

                }
            }
        ]
    }
    ```

1. Modify the `Principal` section of the pasted code snippet accordingly:

>     1. Ensure that this is set to `Federated`, and that the `oidc-provider` setting references the **Provider URL** you [configured above](#step-1-set-up-an-oidc-provider-in-your-aws-account) (that is, `agent.buildkite.com`).

Author: ```suggestion
    1. Ensure that this is set to `Federated`, and points to the `oidc-provider` Amazon Resource Name (ARN) from the **Provider URL** you [configured above](#step-1-set-up-an-oidc-provider-in-your-aws-account) (that is, `agent.buildkite.com`).
```

Author: For clarity, `arn:aws:iam::AWS_ACCOUNT_ID:oidc-provider/agent.buildkite.com` is the full ARN here, and it's a unique reference within AWS (all AWS resources have one).

    1. Change `AWS_ACCOUNT_ID` to your actual AWS account ID.

1. Modify the `Condition` section of the code snippet accordingly:
    1. Ensure the `StringEquals` subsection's _audience_ field name (your provider URL appended by `:aud`—`agent.buildkite.com:aud`) has a value that matches the **Audience** you [configured above](#step-1-set-up-an-oidc-provider-in-your-aws-account) (that is, `sts.amazonaws.com`).

>     1. Ensure the `StringLike` subsection's _subject_ field name (your provider URL appended by `:sub`—`agent.buildkite.com:sub`) has at least one value that matches the format: `organization:ORGANIZATION_SLUG:pipeline:PIPELINE_SLUG:ref:REF:commit:BUILD_COMMIT:step:STEP_KEY`, where the constituent fields of this line determine the conditions (that is, when the values specified in these constituent fields have been met) under which the IAM role is granted in exchange for the OIDC token. The following constituent field's value:

Author: Oh I love how you were able to rephrase this! Thank you!

Reviewer: You're welcome! Thanks for answering all my earlier questions, which certainly helped :-)

            * From the end of your Buildkite URL, after accessing **Pipelines** in the global navigation of your organization in Buildkite.

            * By running the [List organizations](/docs/apis/rest-api/organizations#list-organizations) REST API query to obtain this value from `slug` in the response. For example:

                ```bash
                curl - X GET "https://api.buildkite.com/v2/organizations" \
                  -H "Authorization: Bearer $TOKEN"
                ```
        - `PIPELINE_SLUG` can be obtained:

            * From the end of your Buildkite URL, after accessing **Pipelines** in the global navigation of your organization in Buildkite, then accessing the specific pipeline to be specified in the custom trust policy.

            * By running the [List pipelines](/docs/apis/rest-api/pipelines#list-pipelines) REST API query to obtain this value from `slug` in the response from the specific pipeline. For example:

                ```bash
                curl - X GET "https://api.buildkite.com/v2/organizations/{org.slug}/pipelines" \
                  -H "Authorization: Bearer $TOKEN"
                ```

>         - `REF` is usually replaced with `refs/heads/main` to enforce the IAM role's access and use to only the `main` branch, `refs/tags/*` to ensure only tagged releases are able to be deployed, or a wildcard `*` if the IAM role can access and use all branches.

Author: ```suggestion
        - `REF` is usually replaced with `refs/heads/main` to enforce the IAM role's access and use to only the `main` branch, `refs/tags/*` to ensure only tagged releases are able to be deployed, or a wildcard `*` if the IAM role can be accessed and used by all branches.
```

        - `BUILD_COMMIT` (optional) can be omitted and if so, is usually replaced with a single wildcard `*` at the end of the line.
        - `STEP_KEY` (optional) can be omitted and if so, is usually replaced with a single wildcard `*` at the end of the line.

    **Note:** When formulating your _subject_ field's value, you can replace any of the constituent field values above with a wildcard `*` to not set limits on those constituent fields.

    You can also allow this IAM role to be used with other pipelines, branches, commits and steps by specifying multiple comma-separated values for the `agent.buildkite.com:sub` _subject_ field.

>     For example, to allow the IAM role to be used for any pipeline in the Buildkite organization `example-org`, when building on any branch or tag, specify a single _subject_ field value of `organization:example-org:pipeline:*:ref:refs:*`.

Author: CC @tessereth can you confirm this is valid for our public docs here?

Reviewer 2: Yep, it's correct, though it can be even simpler:

```suggestion
    For example, to allow the IAM role to be used for any pipeline in the Buildkite organization `example-org`, when building on any branch or tag, specify a single _subject_ field value of `organization:example-org:*`.
```

Reviewer: Thanks @tessereth , I'll amend the docs accordingly :-)

1. Modify the `Condition` section's `IpAddress` values (`AGENT_PUBLIC_IP_ONE` and `AGENT_PUBLIC_IP_TWO`) with a list of your agent's IP addresses.

    Only OIDC token exchange requests (for IAM roles) from Buildkite Agents with these IP addresses will be permitted.

1. Verify that your custom trust policy is complete. The following example trust policy (noting that `AWS_ACCOUNT_ID` has not been specified) will only allow the exchange of an agent's OIDC tokens with IAM roles when:

>     * the Buildkite organization is `example-org`

Author: ```suggestion
    * the Buildkite organization is `example-org`
    * the Buildkite pipeline is `example-pipeline`
```

Reviewer: Thanks! 🤦

    * building on both the `main` branch and tagged releases
    * on Buildkite Agents whose IP addresses are either `192.0.2.0` or `198.51.100.0`

    ```json
    {
        "Version": "2012-10-17",
        "Statement": [
            {
                "Effect": "Allow",
                "Principal": {
                    "Federated": "arn\:aws\:iam:\:AWS_ACCOUNT_ID\:oidc-provider/agent.buildkite.com"
                },
                "Action": "sts:AssumeRoleWithWebIdentity",
                "Condition": {
                    "StringEquals": {
                        "agent.buildkite.com:aud": "sts.amazonaws.com"
                    },

>                     "StringLike": {

Reviewer 3: Does this need to be `ForAnyValue:StringLike`? Or is that if the *target* contains multiple keys?

Author: I _think_ that's just when there's multiple keys but now I'm not sure...

Reviewer: Let's wait for more feedback on this one once the published docs are in use. We can always amend this at any point in the future if we need to.

                        "agent.buildkite.com:sub": [
                            "organization\:example-org\:pipeline\:example-pipeline\:ref:refs/heads/main:*",
                            "organization\:example-org\:pipeline\:example-pipeline\:ref:refs/tags/*:*"
                        ]
                    },
                    "IpAddress": {
                        "aws:SourceIp": [
                            "192.0.2.0",
                            "198.51.100.0"
                        ]
                    }

                }
            }
        ]
    }
    ```

> 1. In the **Custom trust policy** section, copy your modified custom trust policy, paste it into your IAM role, and complete the next few steps up to specifying the **Role name**.

Author: ```suggestion
1. In the **Custom trust policy** section, copy your modified custom trust policy, paste it into your IAM role, and complete the next few steps up to specify the **Role name**.
```

Reviewer: From a grammatical perspective, I think 'specify' should be kept in its gerund form.

> 1. Specify an appropriate **Role name**, for example, `compute-ssm-oidc-example`, and complete the remaining steps.

Author: We could maybe replace the `compute` value in the example names here and on lines 161, 172, 200 and 203.

Reviewer: Great idea! Sorry, I hadn't quite got that far

Author: ```suggestion
1. Specify an appropriate **Role name**, for example, `example-pipeline-oidc-for-ssm`, and complete the remaining steps.
```

                ```bash
                curl - X GET "https://api.buildkite.com/v2/organizations/{org.slug}/pipelines" \
                  -H "Authorization: Bearer $TOKEN"
                ```
        - `REF` is usually replaced with `refs/heads/main` to enforce the IAM role's access and use to only the `main` branch, `refs/tags/*` to ensure only tagged releases are able to be deployed, or a wildcard `*` if the IAM role can be accessed and used by all branches.
        - `BUILD_COMMIT` (optional) can be omitted and if so, is usually replaced with a single wildcard `*` at the end of the line.
        - `STEP_KEY` (optional) can be omitted and if so, is usually replaced with a single wildcard `*` at the end of the line.

    **Note:** When formulating your _subject_ field's value, you can replace any of the constituent field values above with a wildcard `*` to not set limits on those constituent fields.

    For example, to allow the IAM role to be used for any pipeline in the Buildkite organization `example-org`, when building on any branch or tag, specify a single _subject_ field value of `organization:example-org:*`.

>     You can also allow this IAM role to be used with other pipelines, branches, commits and steps by specifying multiple comma-separated values for the `agent.buildkite.com:sub` _subject_ field.

Reviewer 4: It's probably worth calling you to the customer that they should **never** omit the organization slug

Reviewer: Thanks @patrobinson , my next revision has specified which constituent field values are optional and which are required.

Would it be OK to specify `organization:*` though? (If not, I'll clarify this.)

Author: While you technically could specify that any organization can assume the IAM role with that approach, you absolutely **should not**, because that'll remove all the protections we're encouraging people to set here. They should ensure that only organisations that they control can assume the role by explicitly naming them here.

Reviewer: Thanks @duckalini , I'll keep the latest guidance as is.

FYI, our docs IA is a bit of a mess and so I'm restructuring the Security section of the Pipelines docs into their own consistent directory and nav hierarchy before moving the OIDC docs into them.

> 1. Modify the `Condition` section's `IpAddress` values (`AGENT_PUBLIC_IP_ONE` and `AGENT_PUBLIC_IP_TWO`) with a list of your agent's IP addresses or CIDRs.

Reviewer 4: I think we should expand on why this conditional is useful:

> Restricting role assumption to specific IP addresses provides defence in depth against an attacker stealing an OIDC token and using it to access your cloud environment.

>     Only OIDC token exchange requests (for IAM roles) from Buildkite Agents with these IP addresses will be permitted.

Reviewer 4: It's worth pointing out this only works if you have dedicated public IPs. One of the options our elastic stack gives is to run agents in public subnets with ephemeral IPs.
We could cross link to the elastic-stack documentation to let them know how to do this.

Reviewer: Need to incorporate relevant static IP configuration content in the relevant pages within https://buildkite.com/docs/agent/v3/elastic-ci-aws/elastic-ci-stack-overview first before we can mention how to configure dedicated IPs in elastic stack.

                        "agent.buildkite.com:sub": [
                            "organization\:example-org\:pipeline\:example-pipeline\:ref:refs/heads/main:*",
                            "organization\:example-org\:pipeline\:example-pipeline\:ref:refs/tags/*:*"
                        ]
                    },
                    "IpAddress": {
                        "aws:SourceIp": [
                            "192.0.2.0",
                            "198.51.100.0"
                        ]
                    }

                }
            }
        ]
    }
    ```

1. In the **Custom trust policy** section, copy your modified custom trust policy, paste it into your IAM role, and complete the next few steps up to specifying the **Role name**.

1. Specify an appropriate **Role name**, for example, `compute-ssm-oidc-example`, and complete the remaining steps.

## Step 3: Configure your IAM role with AWS actions

Add an inline or managed IAM policy to the role to allow the IAM role to perform any actions your pipeline needs. Common examples are permissions to read secrets from SSM and push images to ECR, but this entirely depends on the purpose of your Pipeline.

> In this example we’ll allow access to read an SSM Parameter Store key named `/pipelines-compute/oidc/example-deploy-key` by attaching the following inline policy.

Author: ```suggestion
In this example we’ll allow access to read an SSM Parameter Store key named `/pipelines/example-pipeline/oidc-for-ssm/example-deploy-key` by attaching the following inline policy.
```

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "ssm:GetParameters"
            ],

>             "Resource": "arn\:aws\:ssm\:us-east-1\:012345678910\:parameter/pipelines-compute/oidc/example-deploy-key"

Author: ```suggestion
            "Resource": "arn\:aws\:ssm\:us-east-1\:012345678910\:parameter/pipelines/example-pipeline/oidc-for-ssm/example-deploy-key"
```

        }
    ]
}
```

## Step 4: Configure your pipeline to assume the role

We’ll use two Buildkite Plugins to use the IAM role and to pull in the SSM parameter

- [aws-assume-role-with-web-identity-buildkite-plugin](https://github.com/buildkite-plugins/aws-assume-role-with-web-identity-buildkite-plugin)
- [aws-ssm-buildkite-plugin](https://github.com/buildkite-plugins/aws-ssm-buildkite-plugin)

This can be added to your Pipeline as follows

```yaml
agents:
  queue: mac-small

steps:
 -  label: "\:aws\: Deploy to Production"
    key: deploy-to-production
    command: echo "Example Deploy Key equals \$EXAMPLE_DEPLOY_KEY"
    env:
      AWS_DEFAULT_REGION: us-east-1
      AWS_REGION: us-east-1
    plugins:
      - aws-assume-role-with-web-identity#v1.0.0:

>           role-arn: arn\:aws\:iam::012345678910:role/compute-ssm-oidc-example

Author: ```suggestion
          role-arn: arn\:aws\:iam::012345678910:role/example-pipeline-oidc-for-ssm
```

      - aws-ssm#v1.0.0:
          parameters:

>             EXAMPLE_DEPLOY_KEY: /pipelines-compute/oidc/example-deploy-key

Author: ```suggestion
            EXAMPLE_DEPLOY_KEY: /pipelines/example-pipeline/oidc-for-ssm/example-deploy-key
```
